enhancements/windows-containers/windows-node-egress-proxy.md

```diff
@@ -0,0 +1,347 @@
+proxy, which could affect inter-pod communication, existing application availablity, and security. Also, the pause image
+may not be able to be fetched, preventing new Windows nodes from being bootstrapped. This would require manual
+intervention from the cluster admin or a new release fixing whatever bug is causing the problem.
+
+## Design Details
+
+### Configuring Proxy Environment Variables
+
+As it stands today, the source of truth for cluster-wide proxy settings is the `Proxy` resource with the name `cluster`.
+The contents of the resource are both user-defined, as well as adjusted by the cluster network operator (CNO). Some
+platforms require instances to access certain endpoints to retreive metadata for bootstrapping. CNO has logic to inject
+additional `no-proxy` entries such as `169.254.169.254` and `.${REGION}.compute.internal` into the `Proxy` resource.
+
+OLM is a subscriber to these `Proxy` settings -- it forwards the settings to the CSV of managed operators, so the WMCO
+container will automatically get the required `NO_PROXY`, `HTTP_PROXY`, and `HTTPS_PROXY` environment variables on startup.
+In fact, OLM will update and restart the operator pod with proper environment variables if the `Proxy` resource changes.
+
+On the Linux side this is achieved through `MachineConfigs` -- when global proxy settings are changed, the ignition
+config is re-rendered by MCO (for whom proxy info is injected by CVO) to include a systemd unit 
+`EnvironmentFile=/etc/mco/proxy.env`, with proxy vars set in this file.
+
+WMCO will retrieve the proxy variables by watching the `rendered-worker` `MachineConfig` for changes and parse info from
+the `proxy.env` file. This could be an expensive operation, but since we already parse ignition from the `MachineConfig`
+as part of node bootstrap, the technical cost and complexity is reduced.
+
+Then, once it has the required values, the operator will set the 3 proxy environment variables on the Windows instance
+during node configuration/reconciling through Powershell. Since Windows services read environment from the system 
+registry and the registry does not update until a shutdown/startup cycle, this will require rebooting the instance.
+```powershell
+$Env:<variable-name> = "<new-value>"
+```
+We may have to separately enable [Powershell to use the proxy vars](https://martin.hoppenheit.info/blog/2015/using-powershell-behind-proxy/).
+
+
+### Configuring Custom Trusted Certificates
+
+1. Add the trusted CA injection request ConfigMap to WMCO's manifests directory. 
+   This is a resource that will be updated by CNO based on the content of the global `Proxy` resource.
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  annotations:
+    release.openshift.io/create-only: "true"
+  labels:
+    config.openshift.io/inject-trusted-cabundle: "true"
+  name: trusted-ca
+  namespace: openshift-windows-machine-config-operator
```

Comment:
Would this procedure be same for the third scenario mentioned above 'modifying certificates present in their trustedCA ConfigMap'?
For my understanding, from where do we get the trusted injection request for an updated certificate, is that also from OLM?

Reply:
Yes it'd be the same procedure. [CNO injects the updated certs into this configmap](https://docs.openshift.com/container-platform/4.12/operators/admin/olm-configuring-proxy-support.html#olm-inject-custom-ca_olm-configuring-proxy-support) so we can use it as the source of trust